vpr/src/backend/physical_netlist_writer.cpp

```diff
@@ -0,0 +1,732 @@
+    t_route_segment pip;
+    pip.type = type;
+    pip.tile = tile;
+    pip.wire0 = wire0;
+    pip.wire1 = wire1;
+
+    return pip;
+}
+
+InterchangeWriterVisitor::InterchangeWriterVisitor(vtr::string_internment* strings, Device::Reader& dr, PhysicalNetlist::PhysNetlist::Builder& phys_netlist)
+    : strings_(strings)
+    , dr_(dr)
+    , phys_netlist_(phys_netlist)
+    , pb_graph_pin_lookup_from_index_by_type_(g_vpr_ctx.device().logical_block_types)
+    , ic_primitive(strings->intern_string(vtr::string_view("ic_primitive")))
+    , ic_routing(strings->intern_string(vtr::string_view("ic_routing")))
+    , ic_site_pip(strings->intern_string(vtr::string_view("ic_site_pip")))
+    , ic_generic(strings->intern_string(vtr::string_view("ic_generic")))
+    , ic_gridloc(strings->intern_string(vtr::string_view("ic_gridloc"))) {
+    auto func = std::bind(&InterchangeWriterVisitor::str, this, std::placeholders::_1);
+    process_cell_bel_mappings(dr_, bel_cell_mappings_, func);
+    process_site_pads();
+}
+
+t_route_branch& InterchangeWriterVisitor::walk_route_tree(const t_rt_node* root, t_route_branch& parent_branch) {
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& rr_graph = device_ctx.rr_graph;
+
+    auto rr_node = RRNodeId(root->inode);
+    t_rr_type rr_type = rr_graph.node_type(rr_node);
+    int x = rr_graph.node_xlow(rr_node);
+    int y = rr_graph.node_ylow(rr_node);
+
+    auto& grid_loc = device_ctx.grid[x][y];
+    auto physical_tile = grid_loc.type;
+
+    if (rr_type == IPIN || rr_type == OPIN) {
+        int pin_num = rr_graph.node_pin_num(rr_node);
+        int xoffset = device_ctx.grid[x][y].width_offset;
+        int yoffset = device_ctx.grid[x][y].height_offset;
+        int sub_tile_offset = physical_tile->get_sub_tile_loc_from_pin(pin_num);
+
+        ClusterBlockId iblock = place_ctx.grid_blocks[x - xoffset][y - yoffset].blocks[sub_tile_offset];
+        VTR_ASSERT(iblock);
+        auto pin_name = new_strings_.get_index(get_pb_graph_node_pin_from_block_pin(iblock, pin_num)->port->name);
+
+        // Get site name
+        VTR_ASSERT(grid_loc.meta != nullptr && grid_loc.meta->has(ic_gridloc));
+
+        auto* value = grid_loc.meta->get(ic_gridloc);
+        VTR_ASSERT(value != nullptr);
+
+        auto sites_info = vtr::split(value->front().as_string().get(strings_), ",");
+        VTR_ASSERT(sub_tile_offset < (int)sites_info.size());
+
+        auto site_info = vtr::split(sites_info[sub_tile_offset], ":");
+        auto site_name = new_strings_.get_index(str(std::stoi(site_info[0])));
+
+        auto segment = get_site_pin_segment(SITE_PIN, site_name, pin_name);
+        if (rr_type == IPIN) {
+            t_route_branch new_branch;
+            new_branch.segment = segment;
+
+            parent_branch.branches.push_back(new_branch);
+        } else
+            parent_branch.segment = segment;
+    }
+
```

Comment:
Ok, I might need a bit more time to carefully read through this, but I think the code could be splitted into two functions at tthis very point. One for preparing the parent branch, another one for walking.

Reply:
Separated the prepare parent branch in a different function and split it from the rest